File: 绿盟全国赛决赛参考答案/Python/第一阶段/step2/Task.py
# -*- coding: utf-8 -*-
class Task:

    def virusArea(self,n,m,area):
    #********* Begin *********#
        if area is None: return None
        if area == []: return 0
        # 当数组不为空时，计算行数和列数
        self.c = n
        self.d = m
        x = 0
        for i in range(self.c):
            for j in range(self.d):
                if area[i][j] == 'o':
                    x += 1
                    area = self.change(area, i, j)
        return x

    def change(self, area, i, j):
        area[i][j] = 0
        if i > 0 and area[i - 1][j] == 'o':
            # 置当前点上边的点为0
            area = self.change(area, i - 1, j)
        if i < self.c - 1 and area[i + 1][j] == 'o':
            # 置当前点下边的点为0
            area = self.change(area, i + 1, j)

        if j < self.d - 1 and area[i][j + 1] == 'o':
            # 置当前点右方的点为0
            area = self.change(area, i, j + 1)
        if j > 0 and area[i][j - 1] == 'o':
            # 置当前点左方的点为0
            area = self.change(area, i, j - 1)
        return area
	
	#********* End *********#

File: 绿盟全国赛决赛参考答案/Java/第一阶段/step2/Task.java
package step2;

public class Task {
    
	public int virusArea(int n,int m,char[][] area){
		/********* Begin *********/
		if (area == null || area.length == 0 
				|| area[0].length == 0)
			return 0;

		int count = 0;
		for (int i = 0; i < n; i++)
			for (int j = 0; j < m; j++)

				if (area[i][j] == 'o') {
					count++;
					change(area, i, j, n, m);
				}
		return count++;
		/********* End *********/	
	}

	private void change(char[][] area, 
			int i, int j, int n, int m) {
		if (i < 0 || i >= n || j < 0 || j >= m)
			return;
		if (area[i][j] != 'o')
			return;

		area[i][j] = '0';
		change(area, i + 1, j, n, m);
		change(area, i - 1, j, n, m);
		change(area, i, j + 1, n, m);
		change(area, i, j - 1, n, m);
 
	}
}

